xmtp_mls/migrations/2023-10-29-205333_state_machine_init/up.sql

```diff
@@ -0,0 +1,72 @@
+CREATE TABLE groups (
+    -- Random ID generated by group creator
+    "id" BLOB PRIMARY KEY NOT NULL,
+    -- Based on the timestamp of the welcome message
+    "created_at_ns" BIGINT NOT NULL,
+    -- Enum of GROUP_MEMBERSHIP_STATE
+    "membership_state" INT NOT NULL
+);
+
+-- Allow for efficient sorting of groups
+CREATE INDEX groups_created_at_idx ON groups(created_at_ns);
+
+CREATE INDEX groups_membership_state ON groups(membership_state);
+
+-- Successfully processed messages meant to be returned to the user
+CREATE TABLE group_messages (
+    "id" BLOB PRIMARY KEY NOT NULL,
+    -- Derived via SHA256(CONCAT(decrypted_message_bytes, conversation_id, timestamp))
+    "group_id" BLOB NOT NULL,
+    -- Message contents after decryption
+    "decrypted_message_bytes" BLOB NOT NULL,
+    -- Based on the timestamp of the message
+    "sent_at_ns" BIGINT NOT NULL,
+    -- Enum GROUP_MESSAGE_KIND
+    "kind" INT NOT NULL,
+    -- Could remove this if we added a table mapping installation_ids to wallet addresses
+    "sender_installation_id" BLOB NOT NULL,
+    "sender_wallet_address" TEXT NOT NULL,
+    FOREIGN KEY (group_id) REFERENCES groups(id)
+);
+
+CREATE INDEX group_messages_group_id_sort_idx ON group_messages(group_id, sent_at_ns);
+
+-- Used to keep track of the last seen message timestamp in a topic
+CREATE TABLE topic_refresh_state (
+    "topic" TEXT PRIMARY KEY NOT NULL,
+    "last_message_timestamp_ns" BIGINT NOT NULL
+);
+
+-- This table is required to retry messages that do not send successfully due to epoch conflicts
+CREATE TABLE group_intents (
+    -- Serial ID auto-generated by the DB
+    "id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL,
+    -- Enum INTENT_KIND
+    "kind" INT NOT NULL,
+    "group_id" BLOB NOT NULL,
+    -- Some sort of serializable blob that can be used to re-try the message if the first attempt failed due to conflict
+    "data" BLOB NOT NULL,
+    -- INTENT_STATE,
+    "state" INT NOT NULL,
+    -- The hash of the encrypted, concrete, form of the message if it was published.
+    "message_hash" BLOB,
+    FOREIGN KEY (group_id) REFERENCES groups(id)
+);
+
+CREATE INDEX group_intents_group_id_id ON group_intents(group_id, id);
+
+CREATE TABLE outbound_welcome_messages (
```

Comment:
1. Do we need this as a separate table, as opposed to `data` on `group_intents`?
2. Can this be created on-demand as opposed to pre-created in the beginning?

Reply:
I'm mostly basing this off of the APIs provided by OpenMLS today. There may be ways of reconfiguring OpenMLS to work the way you propose.

I'm suggesting the flow works as such:
1. Create an intent to add members, with just a list of `installation_id`s as the intent data
2. When processing the intent (after having processed all previous intents) we call `group.add_members(...)` with the key packages of all the new members. This will give us back the commit and the welcome message that has to be sent to all members. It also saves a pending commit in the MLS store. The welcome message doesn't get generated anywhere else in OpenMLS.
3. We read the group state from the network and see if the commit will be accepted or rejected. If accepted we merge the pending commit stored in the MLS state.
4. If accepted, we send the welcome messages created in step 2

To do it on demand, we'd need to somehow generate the welcome messages in step 3 or 4. But based on my reading of the OpenMLS docs, it has to happen at Step 2. That's why I'm suggesting a separate table. Alternatively, we could have a mutable field on the intent (maybe called `meta` or something) that we could write the welcome message data to in step 2. That would remove the need for a separate table, but would make more of the intent mutable.

Reply:
Gotcha, so we can't create them on-demand. Yes, I think it might be better to use a `data` or `meta` or `post_commit_data` field on `group_intents` - it creates less complexity in the schema, saves us models and having to marshal them in and out per-field, and saves us having to write separate queries to fetch the `outbound_welcome_messages` later.

Reply:
I'm on board for that